src/conversions/osstr.rs

```diff
@@ -0,0 +1,203 @@
+                // string data
+                py.from_owned_ptr::<PyString>(ffi::PyUnicode_FromWideChar(
+                    wstr.as_ptr(),
+                    wstr.len() as ffi::Py_ssize_t,
+                ))
+                .into()
+            }
+        }
+    }
+}
+
+// There's no FromPyObject implementation for &OsStr because albeit possible on Unix, this would
+// be impossible to implement on Windows. Hence it's omitted entirely
+
+impl FromPyObject<'_> for OsString {
+    fn extract(ob: &PyAny) -> PyResult<Self> {
+        #[cfg(not(windows))]
+        {
+            let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+            // Decode from Python's lossless bytes string representation back into raw bytes
+            let fs_encoded_bytes: &crate::types::PyBytes = unsafe {
+                ob.py()
+                    .from_borrowed_ptr(ffi::PyUnicode_EncodeFSDefault(pystring.as_ptr()))
+            };
+
+            // Create an OsStr view into the raw bytes from Python
+            let os_str: &OsStr =
+                std::os::unix::ffi::OsStrExt::from_bytes(fs_encoded_bytes.as_bytes());
+
+            Ok(os_str.to_os_string())
+        }
+
+        #[cfg(windows)]
+        {
+            let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+            // Take the quick and easy shortcut if UTF-8
+            if let Ok(utf8_string) = pystring.to_str() {
+                return Ok(utf8_string.to_owned().into());
+            }
+
+            // Get an owned allocated wide char buffer from PyString, which we have to deallocate
+            // ourselves
+            let size =
+                unsafe { ffi::PyUnicode_AsWideChar(pystring.as_ptr(), std::ptr::null_mut(), 0) };
+            if size == -1 {
+                return Err(PyErr::fetch(ob.py()));
+            }
+
+            let mut buffer = vec![0; size as usize];
+            let bytes_read =
+                unsafe { ffi::PyUnicode_AsWideChar(pystring.as_ptr(), buffer.as_mut_ptr(), size) };
+            assert_eq!(bytes_read, size);
+
+            // Copy wide char buffer into OsString
+            let os_string = std::os::windows::ffi::OsStringExt::from_wide(&buffer);
+
+            Ok(os_string)
+        }
+    }
+}
+
+impl IntoPy<PyObject> for &'_ OsStr {
+    #[inline]
+    fn into_py(self, py: Python) -> PyObject {
+        self.to_object(py)
+    }
+}
+
+impl ToPyObject for Cow<'_, OsStr> {
+    #[inline]
+    fn to_object(&self, py: Python) -> PyObject {
+        (&self as &OsStr).to_object(py)
+    }
+}
+
+impl ToPyObject for OsString {
+    #[inline]
+    fn to_object(&self, py: Python) -> PyObject {
+        (&self as &OsStr).to_object(py)
+    }
+}
+
+impl IntoPy<PyObject> for OsString {
+    fn into_py(self, py: Python) -> PyObject {
+        self.to_object(py)
+    }
+}
+
+impl<'a> IntoPy<PyObject> for &'a OsString {
```

Comment:
types/string.rs has an equivalent implementation. I oriented myself at those trait implementations

Reply:
Ahh ok. 

I'm not entirely sure why that impl exists... it might be because `#[pyo3(get)]` on `String` fields needs it? Not sure.

If we can't figure out why we need this impl, I'd rather skip it for now. We can always add it later!